File: app/src/main/java/com/dvor/my/mydvor/NewsFragment.java

package com.dvor.my.mydvor;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.LinkedList;
import java.util.List;

public class NewsFragment extends Fragment
 {

    private static List<MyEventListener> eventListeners;

    public void addEventListener(MyEventListener eventListener) {
        eventListeners.add(eventListener);
    }

    public void notifyEventListeners(MyEvent event) {
        for (MyEventListener eventListener: eventListeners) {
            eventListener.processEvent(event);
        }
    }

    private FirebaseAuth mAuth;
    String userStreetId;
    String organizationId;
    String userBuildingId;
    String userName;
    DatabaseReference myRef2;
    DatabaseReference myRef3;
    DataSnapshot newsSnapshot;
    ValueEventListener listenerBuilding;
    ValueEventListener listenerNews;
    public static boolean updateUIflag;
    private View view;
    private List<News> news = new ArrayList<News>();
    private List<String> likes = new ArrayList<String>();
    private List<String> dislikes = new ArrayList<String>();
    private List<NewsAdapter.Comment> comments = new ArrayList<NewsAdapter.Comment>();
    private String imgID="newsImages/no";
    ListView newsList;
    AdapterView.OnItemClickListener itemListener;
    Context context;
    private TextView postText;
    Uri imageUri;
    public  static final int RESULT_GALLERY = 0;
    int lastID=0;
    int index;
    int top;

     //сохранение текста неотправленного сообщения
     //костыльно но работает
     public void onSaveInstanceState(Bundle outState) {
         outState.putString("postText", (postText.getText().toString()));
         outState.putString("postImg", imgID);
         MainActivity.savedPost=postText.getText().toString();
         MainActivity.postImg=imgID;
     }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        updateUIflag = true;

        if(eventListeners == null) {
            eventListeners = new LinkedList<>();
        }
        else {
            eventListeners.clear();
        }

        view = inflater.inflate(R.layout.fragment_news, container, false);
        // начальная инициализация списка
        // получаем элемент ListView
        newsList = (ListView) view.findViewById(R.id.newsList);

        // создаем адаптер
        context = view.getContext();

        Button button = (Button) view.findViewById(R.id.send_post);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                switch (v.getId()) {
                    case R.id.send_post:
                        sendPost();
                        break;
                    default:
                        break;
                }
            }
        });

        ImageButton imgbutton = (ImageButton) view.findViewById(R.id.add_img);
        imgbutton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                switch (v.getId()) {
                    case R.id.add_img:
                        addImage();
                        break;
                    case R.id.imgPref: {
                        deletePrefImg();
                        break;
                    }
                    default:
                        break;
                }
            }
        });

        ImageButton imgPref = (ImageButton) view.findViewById(R.id.imgPref);
        BitmapDrawable preImg = new BitmapDrawable(context.getResources(),MainActivity.imgPref);
        imgPref.setBackground(preImg);
        imgPref.setOnClickListener(new View.OnClickListener() {
            @Override
           public void onClick(View v) {
              switch (v.getId()) {
                  case R.id.imgPref: {
                       deletePrefImg();
                    }
                       break;
                    default:
                        break;
               }
            }
        });
        if (MainActivity.imgPref!=null)
            imgPref.setImageResource(R.drawable.delete);


        postText = (TextView)view.findViewById(R.id.post_text) ;
        mAuth = FirebaseAuth.getInstance();

        postText.setText(MainActivity.savedPost);
        DatabaseReference myRef = FirebaseDatabase.getInstance().getReference("users").child(mAuth.getUid());
        imgID=MainActivity.postImg;

        myRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if(dataSnapshot.getValue() != null) {
                    userName=dataSnapshot.child("name").getValue().toString()
                    +" "+dataSnapshot.child("surname").getValue().toString();
                    userStreetId = dataSnapshot.child("street_id").getValue().toString();
                    userBuildingId = dataSnapshot.child("building_id").getValue().toString();

                    notifyEventListeners(new MyEvent(this, MyEvent.Type.UpdateAddressID));
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

        this.addEventListener(new MyEventListener() {
            @Override
            public void processEvent(MyEvent event) {
                if(event.getSource() == null || event.getType() == null) {
                    return;
                }

                switch (event.getType()) {
                    case UpdateAddressID: {

                        if(myRef2 != null) {
                            myRef2.removeEventListener(listenerBuilding);
                        }

                        myRef2 = FirebaseDatabase.getInstance().getReference("streets").child(userStreetId).child("buildings").child(userBuildingId).child("organization_id");

                        listenerBuilding = new ValueEventListener() {
                            @Override
                            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                organizationId = dataSnapshot.getValue().toString();

                                notifyEventListeners(new MyEvent(this, MyEvent.Type.UpdateOrganizationId));
                            }

                            @Override
                            public void onCancelled(@NonNull DatabaseError databaseError) {

                            }
                        };

                        myRef2.addValueEventListener(listenerBuilding);
                    } break;

                    case UpdateOrganizationId: {

                        if(myRef3 != null) {
                            myRef3.removeEventListener(listenerNews);
                        }

                        myRef3 = FirebaseDatabase.getInstance().getReference("organization").child(organizationId).child("news");

                        listenerNews = new ValueEventListener() {
                            @Override
                            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                newsSnapshot = dataSnapshot;

                                notifyEventListeners(new MyEvent(this, MyEvent.Type.UpdateNews));
                            }

                            @Override
                            public void onCancelled(@NonNull DatabaseError databaseError) {

                            }
                        };

                        myRef3.addValueEventListener(listenerNews);
                    } break;

                    case UpdateNews: {
                        if(updateUIflag) {
                            updateUI();
                            System.gc();
                        } else {
                            updateUIflag = true;
                        }
                    } break;
                }

            }
        });

        return view;
    }

    private void deletePrefImg()
    {
        MainActivity.imgPref = null;
        imgID="newsImages/no";
        ImageButton imgPref = (ImageButton) view.findViewById(R.id.imgPref);
        imgPref.setBackground(null);
        MainActivity.postImg="newsImages/no";

        imgPref.setImageResource(R.drawable.transparent);
    }
    private void addImage(){
        imgID ="newsImages/"+Integer.toString(lastID+1)+".jpg";
         final int RESULT_LOAD_IMAGE = 1;
         Intent galleryIntent = new Intent(
                Intent.ACTION_PICK,
                android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
        startActivityForResult(galleryIntent , RESULT_GALLERY );
    }

    private void sendPost()
    {
        int newsId=lastID+1;

        if (postText.getText().toString()==null)
           return;
        if (postText.getText().toString().trim().length() == 0)
           return;

       NewsBD mes=new NewsBD(userName, postText.getText().toString(),
               (new Date()).toString(), imgID, mAuth.getUid());

       postText.setText("");
       MainActivity.savedPost="";
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
       FirebaseDatabase.getInstance().getReference("organization")
              .child(organizationId).child("news")
               .child(Integer.toString(newsId)).setValue(mes);
       Toast.makeText(getActivity().getApplicationContext(), "Ваша новость опубликована",
               Toast.LENGTH_SHORT).show();
       deletePrefImg();
    }

     @Override
     public void onActivityResult(int requestCode, int resultCode, Intent data) {
         super.onActivityResult(requestCode, resultCode, data);

         switch (requestCode) {
             case RESULT_GALLERY :
                 if (null != data) {

                     imageUri = data.getData();
                     try{
                     final InputStream imageStream = getActivity().getContentResolver().openInputStream(imageUri);
                     final Bitmap selectedImage = BitmapFactory.decodeStream(imageStream);
                     Storage.uploadPicture(selectedImage, imgID);
                     ImageButton imgPref = (ImageButton) view.findViewById(R.id.imgPref);
                     imgPref.setImageResource(R.drawable.delete);}
                     catch (Exception ex)
                     {
                         Log.d("state", ex.getMessage());
                     }
                 }
                 break;
             default:
                 break;
         }
     }

    private void updateUI(){
        // сохраняем видимую позицию listView
        index = newsList.getFirstVisiblePosition();
        View v = newsList.getChildAt(0);
        top = (v == null) ? 0 : (v.getTop() - newsList.getPaddingTop());

        news.clear();
        likes.clear();
        dislikes.clear();
        comments.clear();
        if(newsSnapshot != null) {
            for (DataSnapshot n : newsSnapshot.getChildren()) {

                news.add(new News(n.getKey(),
                        n.child("header").getValue().toString(),
                        n.child("text").getValue().toString(), n.child("date").getValue().toString(),
                        n.child("img").getValue().toString(), n.child("uID").getValue().toString()));

                lastID=Integer.parseInt(n.getKey());
                long countLikes = 0;
                long countDislikes = 0;
                NewsAdapter.Comment comment = NewsAdapter.Comment.absent;

                DataSnapshot likesSnapshot = n .child("likes");
                DataSnapshot dislikesSnapshot = n .child("dislikes");

                if(likesSnapshot.getValue() != null){
                    countLikes = likesSnapshot.getChildrenCount();
                }

                if(dislikesSnapshot.getValue() != null){
                    countDislikes = dislikesSnapshot.getChildrenCount();
                }

                if(likesSnapshot.child(mAuth.getUid()).getValue() != null){
                    comment = NewsAdapter.Comment.like;
                }
                else if(dislikesSnapshot.child(mAuth.getUid()).getValue() != null){
                    comment = NewsAdapter.Comment.dislike;
                }

                likes.add(Long.toString(countLikes));
                dislikes.add(Long.toString(countDislikes));
                comments.add(comment);
            }

        }

        // отображаем сначала новые
        Collections.reverse(news);
        Collections.reverse(likes);
        Collections.reverse(dislikes);
        Collections.reverse(comments);
        NewsAdapter newsAdapter = new NewsAdapter(getActivity(), R.layout.list_news, news, organizationId, likes, dislikes, comments);
        // устанавливаем адаптер
        newsList.setAdapter(newsAdapter);

        if(itemListener == null) {
            // слушатель выбора в списке
            itemListener = new AdapterView.OnItemClickListener() {
                @Override
                public void onItemClick(AdapterView<?> parent, View v, int position, long id) {
                    // получаем выбранный пункт
                    News selectedNews = (News) parent.getItemAtPosition(position);
                    Toast.makeText(getActivity().getApplicationContext(), "Был выбран пункт " + selectedNews.getTitle(),
                            Toast.LENGTH_SHORT).show();
                }
            };
            newsList.setOnItemClickListener(itemListener);
        }
        //перематывает listView на ранее открытую позицию
        newsList.setSelectionFromTop(index, top);
    }

    @Override
    public void onDestroyView()
    {
        super.onDestroyView();
        eventListeners.clear();
        if(myRef2 != null) {
            myRef2.removeEventListener(listenerBuilding);
        }
        if(myRef3 != null) {
            myRef3.removeEventListener(listenerNews);
        }
        listenerBuilding=null;
        listenerNews=null;
    }
}

File: app/src/main/java/com/dvor/my/mydvor/news/NewsFragment.kt
package com.dvor.my.mydvor.news

import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import com.dvor.my.mydvor.*
import com.dvor.my.mydvor.R
import com.dvor.my.mydvor.data.News
import com.dvor.my.mydvor.data.NewsBD
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import java.util.*

class NewsFragment : Fragment() {
    private lateinit var mAuth: FirebaseAuth
    internal var userStreetId: String = ""
    internal var organizationId: String = ""
    internal var userBuildingId: String = ""
    internal var userName: String = ""
    internal var myRef2: DatabaseReference? = null
    internal var myRef3: DatabaseReference? = null
    internal var newsSnapshot: DataSnapshot? = null
    internal var listenerBuilding: ValueEventListener? = null
    internal var listenerNews: ValueEventListener? = null
    private val news = ArrayList<News>()
    private val likes = ArrayList<String>()
    private val dislikes = ArrayList<String>()
    private val comments = ArrayList<NewsAdapter.Comment>()
    private var imgID = "newsImages/no"
    private lateinit var newsList: ListView
    internal lateinit var context: Context
    private var postText: TextView? = null
    private var imageUri: Uri? = null
    private var lastID = 0
    private var index: Int = 0
    private var top: Int = 0

    private fun addEventListener(eventListener: MyEventListener) {
        eventListeners!!.add(eventListener)
    }

    fun notifyEventListeners(event: MyEvent) {
        for (eventListener in eventListeners!!) {
            eventListener.processEvent(event)
        }
    }

    //сохранение текста неотправленного сообщения
    //костыльно но работает
    override fun onSaveInstanceState(outState: Bundle) {
        if (postText == null) {
            outState.putString("postText", MainActivity.savedPost)
        } else {
            outState.putString("postText", postText!!.text.toString())
            MainActivity.savedPost = postText!!.text.toString()
        }

        outState.putString("postImg", imgID)
        MainActivity.postImg = imgID
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        super.onCreate(savedInstanceState)

        updateUIflag = true

        if (eventListeners == null) {
            eventListeners = LinkedList()
        } else {
            eventListeners!!.clear()
        }

        val view = inflater.inflate(R.layout.fragment_news, container, false)
        // начальная инициализация списка
        // получаем элемент ListView
        newsList = view!!.findViewById(R.id.lv_news)

        // создаем адаптер
        context = view.context

        view.findViewById<Button>(R.id.b_send).setOnClickListener { v ->
            if (v.id == R.id.b_send) {
                sendPost()
            }
        }

        view.findViewById<ImageButton>(R.id.im_add_image).setOnClickListener { v ->
            when (v.id) {
                R.id.im_add_image -> addImage()
                R.id.ib_delete_image -> deletePrefImg()
            }
        }

        val deleteBtn = view.findViewById<ImageButton>(R.id.ib_delete_image)
        deleteBtn.setOnClickListener { v ->
            if (v.id == R.id.ib_delete_image) {
                deletePrefImg()
            }
        }
        deleteBtn.visibility = View.VISIBLE

        postText = view.findViewById(R.id.et_post)
        mAuth = FirebaseAuth.getInstance()

        postText!!.text = MainActivity.savedPost
        val myRef = FirebaseDatabase.getInstance().getReference("users").child(mAuth.uid!!)
        imgID = MainActivity.postImg

        myRef.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.value != null) {
                    userName = (dataSnapshot.child("name").value!!.toString()
                            + " " + dataSnapshot.child("surname").value!!.toString())
                    userStreetId = dataSnapshot.child("street_id").value!!.toString()
                    userBuildingId = dataSnapshot.child("building_id").value!!.toString()

                    notifyEventListeners(MyEvent(this, Type.UpdateAddressID))
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })

        this.addEventListener(object : MyEventListener {
            override fun processEvent(event: MyEvent) {
                if (event.source == null) {
                    return
                }

                when (event.type) {
                    Type.UpdateAddressID -> {

                        myRef2 = FirebaseDatabase.getInstance().getReference("streets").child(userStreetId).child("buildings").child(userBuildingId).child("organization_id")

                        listenerBuilding = object : ValueEventListener {
                            override fun onDataChange(dataSnapshot: DataSnapshot) {
                                organizationId = dataSnapshot.value!!.toString()

                                notifyEventListeners(MyEvent(this, Type.UpdateOrganizationId))
                            }

                            override fun onCancelled(databaseError: DatabaseError) {

                            }
                        }

                        myRef2!!.addValueEventListener(listenerBuilding!!)
                    }

                    Type.UpdateOrganizationId -> {

                        myRef3 = FirebaseDatabase.getInstance().getReference("organization").child(organizationId).child("news")

                        listenerNews = object : ValueEventListener {
                            override fun onDataChange(dataSnapshot: DataSnapshot) {
                                newsSnapshot = dataSnapshot

                                notifyEventListeners(MyEvent(this, Type.UpdateNews))
                            }

                            override fun onCancelled(databaseError: DatabaseError) {

                            }
                        }

                        myRef3!!.addValueEventListener(listenerNews!!)
                    }

                    Type.UpdateNews -> {
                        if (updateUIflag) {
                            updateUI()
                            System.gc()
                        } else {
                            updateUIflag = true
                        }
                    }
                }
            }
        })

        return view
    }

    private fun deletePrefImg() {
        MainActivity.imgPref = null
        imgID = "newsImages/no"
        view!!.findViewById<ImageButton>(R.id.ib_delete_image).visibility = View.GONE
        MainActivity.postImg = "newsImages/no"
    }

    private fun addImage() {
        imgID = "newsImages/" + (lastID + 1) + ".jpg"
        val galleryIntent = Intent(
                Intent.ACTION_PICK,
                android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
        startActivityForResult(galleryIntent, RESULT_GALLERY)
    }

    private fun sendPost() {
        val newsId = lastID + 1

        if (postText!!.text.toString().trim { it <= ' ' }.isEmpty())
            return

        val mes = NewsBD(userName, postText!!.text.toString(),
                Date().toString(), imgID, mAuth.uid.toString())

        postText!!.text = ""
        MainActivity.savedPost = ""
        try {
            Thread.sleep(1000)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

        FirebaseDatabase.getInstance().getReference("organization")
                .child(organizationId).child("news")
                .child(newsId.toString()).setValue(mes)
        Toast.makeText(activity!!.applicationContext, "Ваша новость опубликована",
                Toast.LENGTH_SHORT).show()
        deletePrefImg()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == RESULT_GALLERY && data != null) {
            // remember image data
            imageUri = data.data

            try {
                // open stream for image data
                val imageStream = activity!!.contentResolver.openInputStream(imageUri!!)
                // decode this as bitmap
                val selectedImage = BitmapFactory.decodeStream(imageStream)
                // upload picture to firebase
                Storage.uploadPicture(selectedImage, imgID)
                // show `delete uploaded picture` button
                view!!.findViewById<ImageButton>(R.id.ib_delete_image).visibility = View.VISIBLE
            } catch (ex: Exception) {
                Log.d("state", ex.message.toString())
            }
        }
    }

    private fun updateUI() {
        // сохраняем видимую позицию listView
        index = newsList.firstVisiblePosition
        val v = newsList.getChildAt(0)
        top = if (v == null) 0 else v.top - newsList.paddingTop

        news.clear()
        likes.clear()
        dislikes.clear()
        comments.clear()
        if (newsSnapshot != null) {
            for (n in newsSnapshot!!.children) {

                news.add(News(n.key.toString(),
                        n.child("header").value.toString(),
                        n.child("text").value.toString(), n.child("date").value.toString(),
                        n.child("img").value.toString(), n.child("uid").value.toString()))

                var countLikes: Long = 0
                var countDislikes: Long = 0
                var comment: NewsAdapter.Comment = NewsAdapter.Comment.Absent

                val likesSnapshot = n.child("likes")
                val dislikesSnapshot = n.child("dislikes")

                if (likesSnapshot.value != null) {
                    countLikes = likesSnapshot.childrenCount
                }

                if (dislikesSnapshot.value != null) {
                    countDislikes = dislikesSnapshot.childrenCount
                }

                if (likesSnapshot.child(mAuth.uid!!).value != null) {
                    comment = NewsAdapter.Comment.Like
                } else if (dislikesSnapshot.child(mAuth.uid!!).value != null) {
                    comment = NewsAdapter.Comment.Dislike
                }

                likes.add(countLikes.toString())
                dislikes.add(countDislikes.toString())
                comments.add(comment)
            }
            lastID = Integer.parseInt(newsSnapshot!!.children.count().toString())
        }

        // отображаем сначала новые
        news.reverse()
        likes.reverse()
        dislikes.reverse()
        comments.reverse()
        val newsAdapter = NewsAdapter(activity, R.layout.list_news, news, organizationId, likes, dislikes, comments)
        // устанавливаем адаптер
        newsList.adapter = newsAdapter

        //перематывает listView на ранее открытую позицию
        newsList.setSelectionFromTop(index, top)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        eventListeners!!.clear()
        if (myRef2 != null) {
            myRef2!!.removeEventListener(listenerBuilding!!)
        }
        if (myRef3 != null) {
            myRef3!!.removeEventListener(listenerNews!!)
        }
        listenerBuilding = null
        listenerNews = null
    }

    companion object {
        private var eventListeners: MutableList<MyEventListener>? = null
        var updateUIflag: Boolean = false
        const val RESULT_GALLERY = 0
    }
}